honeybadgermpc/progs/fixedpoint.py

```diff
@@ -0,0 +1,248 @@
+    return int(x * 2 ** f)
+
+
+"""
+Convert a number back from fixed point int form to python float.
+"""
+
+
+def from_fixed_point_repr(x, k=K, f=F, signed=True):
+    x = x.value
+    if x >= 2 ** (k - 1) and signed:
+        x = -(p - x)
+
+    return float(x) / 2 ** f
+
+
+def binary_repr(x, k):
+    """
+    Convert x to a k-bit representation
+    Least significant bit first
+    """
+
+    def _binary_repr(v):
+        res = []
+        v = int(v)
+        for i in range(k):
+            res.append(v % 2)
+            v //= 2
+        return res
+
+    assert type(x) is int
+    return _binary_repr(x)
+
+
+# MPC operations for fixed point
+# Get a random integer from [0  ... 2**m -1) range or m bit random number
+async def random2m(ctx, m):
+    result = ctx.Share(0)
+    bits = []
+    for i in range(m):
+        bits.append(ctx.preproc.get_bit(ctx))
+        result = result + Field(2) ** i * bits[-1]
+
+    return result, bits
+
+
+async def trunc_pr(ctx, x, k, m):
+    """
+    k: Maximum number of bits
+    m: Truncation bits
+    """
+    assert k > m
+    r1, _ = await random2m(ctx, m)
+    r2, _ = await random2m(ctx, k + KAPPA - m)
+    r2 = ctx.Share(r2.v * Field(2) ** m)
+    c = await (x + Field(2 ** (k - 1)) + r1.v + r2.v).open()
+    c2 = c.value % (2 ** m)
+    d = ctx.Share((x.v - Field(c2) + r1.v) * ~(Field(2) ** m))
+    return d
+
+
+async def get_carry_bit(ctx, a_bits, b_bits, low_carry_bit=1):
+    a_bits.reverse()
+    b_bits.reverse()
+    assert len(a_bits) == len(b_bits)
+
+    async def _bit_ltl_reduce(x):
+        if len(x) == 1:
+            return x[0]
+        carry1, all_one1 = await _bit_ltl_reduce(x[: len(x) // 2])
+        carry2, all_one2 = await _bit_ltl_reduce(x[len(x) // 2 :])
+        return carry1 + (await (all_one1 * carry2)), (await (all_one1 * all_one2))
+
+    carry_bits = [(await (ai * bi)) for ai, bi in zip(a_bits, b_bits)]
+    all_one_bits = [
+        ctx.Share(ai.v + bi.v - 2 * carryi.v)
+        for ai, bi, carryi in zip(a_bits, b_bits, carry_bits)
+    ]
+    carry_bits.append(ctx.Share(low_carry_bit))
+    all_one_bits.append(ctx.Share(0))
+    return (await _bit_ltl_reduce(list(zip(carry_bits, all_one_bits))))[0]
+
+
+async def bit_ltl(ctx, a, b_bits):
+    """
+    a: Public
+    b: List of private bit shares. Least significant digit first
+    """
+    b_bits = [ctx.Share(Field(1) - bi.v) for bi in b_bits]
+    a_bits = [ctx.Share(ai) for ai in binary_repr(int(a), len(b_bits))]
+
+    carry = await get_carry_bit(ctx, a_bits, b_bits)
+    return ctx.Share(Field(1) - carry.v)
+
+
+async def mod2m(ctx, x, k, m):
+    r1, r1_bits = await random2m(ctx, m)
+    r2, _ = await random2m(ctx, k + KAPPA - m)
+    r2 = ctx.Share(r2.v * Field(2) ** m)
+
+    c = await (x + r2 + r1 + Field(2) ** (k - 1)).open()
+    c2 = int(c) % (2 ** m)
+    u = await bit_ltl(ctx, c2, r1_bits)
+    a2 = ctx.Share(Field(c2) - r1.v + (2 ** m) * u.v)
+    return a2
+
+
+async def trunc(ctx, x, k, m):
```

Comment:
Could we add some comments to these functions explaining their arguments and intended input/output? It's not so self explanatory